lookout/style/format/robustness.py

```diff
@@ -0,0 +1,246 @@
+    X, y, vnodes_y, _ = fe.extract_features(files)
+    return vnodes_y
+
+
+def files2mispreds(files: Iterable[str], rules: Rules, client: str, language: str
+                   ) -> Iterable[Misprediction]:
+    """
+    Return the model's `Mispredictions` on a list of files. 
+
+    :param files: List of files to get `Mispredictions` from.
+    :param rules: rules of the style-analyzer model.
+    :param client: Babelfish client. Babelfish server should be started accordingly.
+    :param language: Language to consider, others will be discarded.
+    :return: List of `Mispredictions` extracted from a given list of files.
+    """
+    files = prepare_files(files, client, language)
+    fe = FeatureExtractor(language=language, **rules.origin_config["feature_extractor"])
+    X, y, vnodes_y, _ = fe.extract_features(files)
+    X, _ = fe.select_features(X, y)
+    y_pred, winner = rules.predict(X, True)
+    mispreds = get_mispreds(y, y_pred, vnodes_y, winner)
+    return mispreds
+
+
+def get_mispreds(y: numpy.ndarray, y_pred: numpy.ndarray, nodes: Iterable[VirtualNode],
+                 winner: numpy.ndarray) -> Iterable[Misprediction]:
+    """
+    Given 2 Numpy 1-dimensional arrays of labels, return the list of `Mispredictions`
+    where the labels differ.
+
+    :param y: Numpy 1-dimensional array of labels.
+    :param y_pred: Numpy 1-dimensional array of predicted labels by the model.
+    :param nodes: List of `VirtualNodes`.
+    :param winner: Numpy 1-dimensional array of the winning rule indices for each sample.
+    :return: List of `Mispredictions` where the labels `y` and `y_pred` differ. 
+    """
+    mispreds = []
+    for gt, pred, vn, rule in zip(y, y_pred, nodes, winner):
+        if gt != pred:
+            mispreds.append(Misprediction(gt, pred, vn, rule))
+    return mispreds
+
+
+def get_diff_mispreds(mispreds: Iterable[Misprediction], lines_changed: Dict[str, Set[int]]
+                      ) -> Dict[str, Misprediction]:
+    """
+.   Filter a list of `Mispredictions` to select only those involving at least one line
+    that has been modified by adding noise.
+
+    :param mispreds: List of `Mispredictions` to filter.
+    :param lines_changed: Dict of lines that have been changed when adding random noise.
+    :return: Dictionary of the `Mispredictions` involving at least one line
+        that has been modified when adding random noise. 
+    """
+    diff_mispreds = {}
+    for m in mispreds:
+        mispred_lines = set(range(m.node.start.line, m.node.end.line + 1))
+        if set.intersection(mispred_lines, lines_changed[m.node.path]):
+            try:
+                if m.node.start.offset < diff_mispreds[m.node.path].node.start.offset:
+                    diff_mispreds[m.node.path] = m
+            except KeyError:
+                diff_mispreds[m.node.path] = m
+    return diff_mispreds
+
+
+def get_style_fixes(mispreds: Dict[str, Misprediction], vnodes: Iterable[VirtualNode],
```

Comment:
arguments should use the broadest type applicable: here Mapping[str, Misprediction] , Iterable[str] and Iterable[str] (if you transform the loop into a zip, otherwise Sequence instead of Iterable).

Reply:
ok sorry I though because in the function i used `true_files[i]`, it should be a list

Reply:
So usually when you use somelist[i] the broadest type is Sequence (you can check the abstract base classes here for reference: https://docs.python.org/3/library/collections.abc.html#collections-abstract-base-classes). Here what I was saying is that you can zip true_files and noisy_files to only need iterables:

    style_fixes = []
    for true_file, noisy_file in zip(true_files, noisy_files):
        try:
            mispred = mispreds[noisy_file]
        except KeyError:
            continue
        for vn in vnodes:
            if vn.path == true_file and vn.start.offset == mispred.node.start.offset:
                if mispred.pred == vn.y :
                    style_fixes.append(mispred)
                break
    return style_fixes